cmd/crank/completion/completion.go

```diff
@@ -0,0 +1,259 @@
+		u.SetGroupVersionKind(schema.GroupVersionKind{
+			Kind:    mapping.GroupVersionKind.Kind,
+			Group:   mapping.GroupVersionKind.Group,
+			Version: mapping.GroupVersionKind.Version,
+		})
+		err = client.List(context.Background(), u)
+		if err != nil {
+			return nil
+		}
+
+		var predictions []string
+		for _, res := range u.Items {
+			if strings.HasPrefix(res.GetName(), a.Last) {
+				predictions = append(predictions, res.GetName())
+			}
+		}
+		return predictions
+	}
+}
+
+func contextPredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			clientcmd.NewDefaultClientConfigLoadingRules(),
+			&clientcmd.ConfigOverrides{},
+		)
+
+		kubeConfig, err := clientConfig.RawConfig()
+		if err != nil {
+			return nil
+		}
+
+		var predictions []string
+		for name := range kubeConfig.Contexts {
+			if strings.HasPrefix(name, a.Last) {
+				predictions = append(predictions, name)
+			}
+		}
+		return predictions
+	}
+}
+
+func namespacePredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		client, err := kubernetesClientset()
+		if err != nil {
+			return nil
+		}
+
+		namespaceList, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return nil
+		}
+
+		var predictions []string
+		for _, ns := range namespaceList.Items {
+			if strings.HasPrefix(ns.GetName(), a.Last) {
+				predictions = append(predictions, ns.GetName())
+			}
+		}
+		return predictions
+	}
+}
+
+func kubernetesClientset() (*kubernetes.Clientset, error) {
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+
+	kubeConfig, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(kubeConfig)
+}
+
+func kubernetesClient(configOverrides *clientcmd.ConfigOverrides) (client.Client, *rest.Config, error) {
+	clientconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		configOverrides,
+	)
+
+	kubeconfig, err := clientconfig.ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client, err := client.New(rest.CopyConfig(kubeconfig), client.Options{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return client, rest.CopyConfig(kubeconfig), nil
+}
+
+func parseConfigOverride(a complete.Args) *clientcmd.ConfigOverrides {
+	context := ""
+	for i, arg := range a.All {
+		if arg == "--context" && i < len(a.All) {
+			context = a.All[i+1]
+		}
+	}
+	return &clientcmd.ConfigOverrides{
+		CurrentContext: context,
+	}
+}
+
+// Copied over from cli-runtime pkg/resource Builder,
+// https://github.com/kubernetes/cli-runtime/blob/9a91d944dd43186c52e0162e12b151b0e460354a/pkg/resource/builder.go#L768
+func mappingFor(rmapper meta.RESTMapper, resourceOrKindArg string) (*meta.RESTMapping, error) {
```

Comment:
As you noted, we've already copied from upstream kubernetes to https://github.com/crossplane/crossplane/blob/release-1.19/cmd/crank/beta/trace/internal/resource/client.go#L44. Instead of copying it again, can we move it to a common location in the cmd where both completions and trace can access it?

Reply:
I agree, that's why I asked for any ideas in this PR. Last time I looked at it, I wasn't able to find a nice spot. Maybe this time :)

Reply:
Okay, I moved it in an internal package for crank.